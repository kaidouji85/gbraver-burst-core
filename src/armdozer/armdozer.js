// @flow

/** アームドーザID */
export type ArmDozerId = string;

// TODO バーストの種類を追加する

/** アームドーザ基本情報 */
export type Armdozer = {
  /** アームドーザID */
  id: ArmDozerId,
  /** 名前 */
  name: string,
  /** 最大HP */
  maxHp: number,
  /** 最大バッテリー */
  maxBattery: number,
  /** 攻撃 */
  power: number,
  /** スピード */
  speed: number,
  /** アームドーザの外見を指定する */
  appearance: string,
};
