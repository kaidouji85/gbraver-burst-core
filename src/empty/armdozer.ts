import type { Armdozer } from "../player/armdozer";
import type { Burst } from "../player/burst";
import type { ArmdozerState } from "../state/armdozer-state";

/** 空のバースト */
export const EMPTY_BURST: Burst = {
  type: "RecoverBattery",
  recoverBattery: 0,
  turnStartBatteryCorrect: 0,
};

/** 空のアームドーザデータ */
export const EMPTY_ARMDOZER: Armdozer = {
  id: "id",
  name: "name",
  maxHp: 3000,
  maxBattery: 5,
  power: 2000,
  speed: 2000,
  burst: EMPTY_BURST,
};

/** 空のアームドーザゲームステート */
export const EMPTY_ARMDOZER_STATE: ArmdozerState = {
  ...EMPTY_ARMDOZER,
  battery: EMPTY_ARMDOZER.maxBattery,
  hp: EMPTY_ARMDOZER.maxHp,
  effects: [],
  enableBurst: true,
};
