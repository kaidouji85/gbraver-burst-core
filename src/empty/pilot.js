// @flow

import type { Pilot } from "../player/pilot";
import type { PilotState } from "../state/pilot-state";

/**
 * 空のパイロット
 */
export const EMPTY_PILOT: Pilot = {
  id: "EMPTY_PLAYER",
  name: "名無し",
  skill: {
    type: "RecoverBatterySkill",
    recoverBattery: 2,
  },
};

/**
 * 空のパイロットステート
 */
export const EMPTY_PILOT_STATE: PilotState = {
  ...EMPTY_PILOT,
  enableSkill: true,
};
