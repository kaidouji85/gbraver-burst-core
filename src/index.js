// @flow
export {ArmDozerIdList, ArmdozerAppearances, ArmDozers} from './master/armdozers';
export {GbraverBurstCore} from './game/index';

export type {Player, PlayerX, PlayerId} from './player/player';
export type {ArmDozerId, ArmdozerAppearance, Armdozer, ArmdozerX} from './player/armdozer';
export type {Burst, RecoverBattery, BuffPower, LightningBarrier, ContinuousAttack} from './player/burst';

export type {GameState, GameStateX} from './game/state/game-state';
export type {PlayerState, PlayerStateX} from './game/state/player-state';
export type {ArmdozerState, ArmdozerStateX} from './game/state/armdozer-state';
export type {ArmdozerEffect, EmptyArmdozerEffect, CorrectPower, TryReflect, ContinuousActivePlayer} from './game/state/armdozer-effect';

export type {Command, PlayerCommand, PlayerCommandX} from './command/command';
export type {BatteryCommand} from './command/battery';
export type {BurstCommand} from './command/burst';
export type {EmptyCommand} from './command/empty-command';

export type {Effect} from './effect/index';
export type {BatteryDeclaration} from './effect/battery-declaration/battery-declaration';
export type {Battle, BattleX} from './effect/battle/battle';
export type {BattleResult} from './effect/battle/result/battle-result';
export type {CriticalHit} from './effect/battle/result/critical-hit';
export type {Feint} from './effect/battle/result/feint';
export type {Guard} from './effect/battle/result/guard';
export type {Miss} from './effect/battle/result/miss';
export type {NormalHit} from './effect/battle/result/normal-hit';
export type {BurstEffect} from './effect/burst/burst-effect';
export type {GameEnd, GameEndX, GameEndResult} from './effect/game-end/game-end';
export type {GameOver, EvenMatch} from './game/end-judging/game-end-judging';
export type {InputCommand, Selectable, NoChoice} from './effect/input-command/input-command';
export type {Reflect, ReflectDamageEffect} from './effect/reflect/reflect';
export type {StartGame} from './effect/start-game/start-game';
export type {TurnChange} from './effect/turn-change/turn-change';
export type {UpdateRemainingTurn, EndArmdozerEffect} from './effect/update-remaning-turn/update-remaining-turn';
export type {RightItself} from './effect/right-itself/right-itself';

// TODO 以下は廃止予定
export {start} from './game/start/start';
export {progress} from './game/progress/index';
