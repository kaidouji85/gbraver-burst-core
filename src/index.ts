export { BatteryCommand, BatteryCommandSchema } from "./command/battery";
export {
  BurstCommand,
  BurstCommandSchema,
} from "./command/burst";
export {
  Command,
  CommandSchema,
  isCommandEqual,
} from "./command/command";
export {
  EmptyCommand,
  EmptyCommandSchema,
} from "./command/empty-command";
export {
  PilotSkillCommand,
  PilotSkillCommandSchema,
} from "./command/pilot-skill";
export { QuickCommand } from "./command/quick-command";
export {
  correctedBattery,
  totalBatteryCorrection,
} from "./effect/battery-correction";
export {
  BatteryDeclaration,
  BatteryDeclarationSchema,
} from "./effect/battery-declaration/battery-declaration";
export {
  Battle,
  BattleSchema,
  BattleX,
} from "./effect/battle/battle";
export { updateDefender } from "./effect/battle/players/update-defender";
export {
  BattleResult,
  battleResult,
  BattleResultSchema,
} from "./effect/battle/result/battle-result";
export {
  CriticalHit,
  CriticalHitSchema,
} from "./effect/battle/result/critical-hit";
export { Feint, FeintSchema } from "./effect/battle/result/feint";
export { Guard, GuardSchema } from "./effect/battle/result/guard";
export { Miss, MissSchema } from "./effect/battle/result/miss";
export {
  NormalHit,
  NormalHitSchema,
} from "./effect/battle/result/normal-hit";
export {
  BurstEffect,
  BurstEffectSchema,
} from "./effect/burst/burst-effect";
export { hasContinuousActive } from "./effect/continuous-active/has-continuous-active";
export {
  correctPower,
  hasHalveCorrectPower,
  totalCorrectPower,
} from "./effect/correct-power";
export {
  GameEnd,
  GameEndResult,
  GameEndResultSchema,
  GameEndX,
} from "./effect/game-end/game-end";
export { Effect } from "./effect/index";
export {
  InputCommand,
  NoChoice,
  Selectable,
} from "./effect/input-command/input-command";
export {
  PilotSkillEffect,
  PilotSkillEffectX,
} from "./effect/pilot-skill/pilot-skill-effect";
export { Reflect } from "./effect/reflect/reflect";
export { RightItself } from "./effect/right-itself/right-itself";
export { StartGame } from "./effect/start-game/start-game";
export { TurnChange } from "./effect/turn-change/turn-change";
export {
  EndArmdozerEffect,
  UpdateRemainingTurn,
} from "./effect/update-remaning-turn/update-remaining-turn";
export {
  EMPTY_ARMDOZER_EFFECT,
  EMPTY_CORRECT_POWER,
} from "./empty/amrdozer-effect";
export {
  EMPTY_ARMDOZER,
  EMPTY_ARMDOZER_STATE,
  EMPTY_BURST,
} from "./empty/armdozer";
export { EMPTY_BATTLE } from "./empty/battle";
export { EMPTY_GAME_STATE } from "./empty/game-state";
export { EMPTY_PILOT, EMPTY_PILOT_STATE } from "./empty/pilot";
export { EMPTY_PLAYER, EMPTY_PLAYER_STATE } from "./empty/player";
export { PlayerCommand, PlayerCommandX } from "./game/command/player-command";
export {
  EvenMatch,
  EvenMatchSchema,
  GameContinue,
  GameContinueSchema,
  GameEndJudging,
  GameEndJudgingSchema,
  GameOver,
  GameOverSchema,
} from "./game/end-judging/game-end-judging";
export {
  GbraverBurstCore,
  RestoreGbraverBurst,
} from "./game/gbraver-burst-core";
export {
  restoreGbraverBurst,
  startGbraverBurst,
} from "./game/gbraver-burst-core";
export { ArmDozerIds, ArmDozers } from "./master/armdozers";
export { PilotIds, Pilots } from "./master/pilots";
export {
  Armdozer,
  ArmDozerId,
  ArmDozerIdSchema,
  ArmdozerSchema,
  ArmdozerX,
} from "./player/armdozer";
export { Burst, BurstSchema } from "./player/burst";
export { BatteryLimitBreak, BatteryLimitBreakSchema } from "./player/burst/battery-limit-break";
export { BuffPower, BuffPowerSchema } from "./player/burst/buff-power";
export { ContinuousAttack, ContinuousAttackSchema } from "./player/burst/continuous-attack";
export { LightningBarrier, LightningBarrierSchema } from "./player/burst/lightning-barrier";
export { RecoverBattery, RecoverBatterySchema } from "./player/burst/recover-battery";
export {
  Pilot,
  PilotId,
  PilotIdSchema,
  PilotSchema,
  PilotX,
} from "./player/pilot";
export { BatteryBoostSkill, BatteryBoostSkillSchema } from "./player/pilot/battery-boost-skill";
export { BatteryEnchantmentSkill, BatteryEnchantmentSkillSchema } from "./player/pilot/battery-enchantment-skill";
export { BuffPowerSkill, BuffPowerSkillSchema } from "./player/pilot/buff-power-skill";
export { DamageHalvedSkill, DamageHalvedSkillSchema } from "./player/pilot/damage-halved-skill";
export { PilotSkill, PilotSkillSchema } from "./player/pilot/pilot-skill";
export { RecoverBatterySkill, RecoverBatterySkillSchema } from "./player/pilot/recover-battery-skill";
export {
  Player,
  PlayerId,
  PlayerIdSchema,
  PlayerSchema,
  PlayerX,
} from "./player/player";
export {
  ArmdozerEffect,
  ArmdozerEffectSchema,
} from "./state/armdozer-effect";
export { BatteryCorrection, BatteryCorrectionSchema } from "./state/armdozer-effect/battery-correction";
export { BatteryRecoverSkip, BatteryRecoverSkipSchema } from "./state/armdozer-effect/battery-recover-skip";
export { ContinuousActivePlayer, ContinuousActivePlayerSchema } from "./state/armdozer-effect/continuous-active-player";
export { CorrectPower, CorrectPowerSchema } from "./state/armdozer-effect/correct-power";
export { DamageHalved, DamageHalvedSchema } from "./state/armdozer-effect/damage-halved";
export { EffectPeriod, EffectPeriodSchema } from "./state/armdozer-effect/effect-period";
export { EmptyArmdozerEffect, EmptyArmdozerEffectSchema } from "./state/armdozer-effect/empty-armdozer-effect";
export { HalveCorrectPower, HalveCorrectPowerSchema } from "./state/armdozer-effect/halve-correct-power";
export { ReflectDamageEffect, ReflectDamageEffectSchema } from "./state/armdozer-effect/reflect-damage-effect";
export { parseSpecialPeriodEffect } from "./state/armdozer-effect/special-period-effect";
export { SpecialPeriodEffectSchema } from "./state/armdozer-effect/special-period-effect";
export { parseTryReflect } from "./state/armdozer-effect/try-reflect";
export { TryReflectSchema } from "./state/armdozer-effect/try-reflect";
export { TryReflect } from "./state/armdozer-effect/try-reflect";
export { parseTurnLimitEffect } from "./state/armdozer-effect/turn-limit-effect";
export { TurnLimitEffectSchema } from "./state/armdozer-effect/turn-limit-effect";
export {
  ArmdozerState,
  ArmdozerStateSchema,
  ArmdozerStateX,
  parseArmDozerState,
} from "./state/armdozer-state";
export { GameState, GameStateX } from "./state/game-state";
export {
  parsePilotState,
  PilotState,
  PilotStateSchema,
  PilotStateX,
} from "./state/pilot-state";
export {
  parsePlayerState,
  PlayerState,
  PlayerStateSchema,
  PlayerStateX,
} from "./state/player-state";
export { isPlayerDeath } from "./state/player-state/is-player-death";
