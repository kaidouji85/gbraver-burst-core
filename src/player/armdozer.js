// @flow

import type { Burst } from "./burst";

/** アームドーザID */
export type ArmDozerId = string;

/**
 * アームドーザ基本情報
 *
 * @typeparam {X} バースト
 */
export type ArmdozerX<X> = {
  /** アームドーザID */
  id: ArmDozerId,
  /** 名前 */
  name: string,
  /** 最大HP */
  maxHp: number,
  /** 最大バッテリー */
  maxBattery: number,
  /** 攻撃 */
  power: number,
  /** スピード */
  speed: number,
  /** バースト */
  burst: X,
};

/** アームドーザ基本情報 */
export type Armdozer = ArmdozerX<Burst>;
