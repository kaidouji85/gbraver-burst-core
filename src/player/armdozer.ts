import { z } from "zod";

import { Burst, BurstSchema } from "./burst";

/** アームドーザID */
export type ArmDozerId = string;

/** ArmDozerId zodスキーマ */
export const ArmDozerIdSchema = z.string();

/**
 * 任意オブジェクトをArmDozerIdにパースする
 * @param origin パース元
 * @return パース結果、パースできない場合はnull
 */
export const parseArmDozerId = (origin: unknown): ArmDozerId | null => {
  const result = ArmDozerIdSchema.safeParse(origin);
  return result.success ? result.data : null;
};

/**
 * アームドーザ基本情報（型指定あり）
 * @template X バースト
 */
export type ArmdozerX<X> = Readonly<{
  /** アームドーザID */
  id: ArmDozerId;
  /** 名前 */
  name: string;
  /** 最大HP */
  maxHp: number;
  /** 最大バッテリー */
  maxBattery: number;
  /** 攻撃 */
  power: number;
  /** スピード */
  speed: number;
  /** バースト */
  burst: X;
}>;

/** アームドーザ基本情報 */
export type Armdozer = ArmdozerX<Burst>;

/** Armdozer zodスキーマ */
export const ArmdozerSchema = z.object({
  id: ArmDozerIdSchema,
  name: z.string(),
  maxHp: z.number(),
  maxBattery: z.number(),
  power: z.number(),
  speed: z.number(),
  burst: BurstSchema,
});

/**
 * 任意オブジェクトをArmdozerにパースする
 * @param origin パース元
 * @return パース結果、パースできない場合はnull
 */
export const parseArmdozer = (origin: unknown): Armdozer | null => {
  const result = ArmdozerSchema.safeParse(origin);
  return result.success ? result.data : null;
};
