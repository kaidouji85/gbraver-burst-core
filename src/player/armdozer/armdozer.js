// @flow

import type {Burst} from "./burst";

/** アームドーザID */
export type ArmDozerId = string;

/** アームドーザ基本情報 */
export type Armdozer = {
  /** アームドーザID */
  id: ArmDozerId,
  /** 名前 */
  name: string,
  /** 最大HP */
  maxHp: number,
  /** 最大バッテリー */
  maxBattery: number,
  /** 攻撃 */
  power: number,
  /** スピード */
  speed: number,
  /** アームドーザの外見を指定する */
  appearance: string,
  /** バースト */
  burst: Burst,
};
