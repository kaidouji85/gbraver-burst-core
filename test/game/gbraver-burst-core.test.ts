import path from "path";

import { PlayerCommand } from "../../src";
import { EMPTY_ARMDOZER } from "../../src/empty/armdozer";
import { EMPTY_PILOT } from "../../src/empty/pilot";
import {
  restoreGBraverBurst,
  startGBraverBurst,
} from "../../src/game";
import type { Player } from "../../src/player/player";
import {
  exportSnapShotJSON,
  importSnapShotJSON,
  shouldUpdateSnapShot,
} from "../snap-shot";

const PLAYER1: Player = {
  playerId: "player1",
  pilot: EMPTY_PILOT,
  armdozer: { ...EMPTY_ARMDOZER, speed: 1600 },
};
const PLAYER2: Player = {
  playerId: "player2",
  pilot: EMPTY_PILOT,
  armdozer: { ...EMPTY_ARMDOZER, speed: 2000 },
};
const COMMAND1: PlayerCommand = {
  playerId: "player1",
  command: {
    type: "BATTERY_COMMAND",
    battery: 3,
  },
};
const COMMAND2: PlayerCommand = {
  playerId: "player2",
  command: {
    type: "BATTERY_COMMAND",
    battery: 2,
  },
};

test("初期状態を正しく作ることができる", () => {
  const core = startGBraverBurst([PLAYER1, PLAYER2]);
  const result = core.stateHistory();
  const snapShotPath = path.join(
    __dirname,
    "gbraver-burst-core__initial-state.json",
  );
  shouldUpdateSnapShot() && exportSnapShotJSON(snapShotPath, result);
  const snapShot = shouldUpdateSnapShot()
    ? result
    : importSnapShotJSON(snapShotPath);
  expect(result).toEqual(snapShot);
});

test("プレイヤー情報が正しくセットされている", () => {
  const core = startGBraverBurst([PLAYER1, PLAYER2]);
  const result = core.players();
  const expected = [PLAYER1, PLAYER2];
  expect(result).toEqual(expected);
});

test("正しくゲームを進めることができる", () => {
  const core = startGBraverBurst([PLAYER1, PLAYER2]);
  const result = core.progress([COMMAND1, COMMAND2]);
  const snapShotPath = path.join(
    __dirname,
    "gbraver-burst-core__progress.json",
  );
  shouldUpdateSnapShot() && exportSnapShotJSON(snapShotPath, result);
  const snapShot = shouldUpdateSnapShot()
    ? result
    : importSnapShotJSON(snapShotPath);
  expect(result).toEqual(snapShot);
});

test("ゲームステート履歴が正しく更新される", () => {
  const core = startGBraverBurst([PLAYER1, PLAYER2]);
  const initialState = core.stateHistory();
  const update = core.progress([COMMAND1, COMMAND2]);
  const result = core.stateHistory();
  const expected = [...initialState, ...update];
  expect(result).toEqual(expected);
});

test("ダンプ、リストアを正しく行うことができる", () => {
  const core = startGBraverBurst([PLAYER1, PLAYER2]);
  core.progress([COMMAND1, COMMAND2]);
  const data = core.dump();
  const restoreCore = restoreGBraverBurst(data);
  expect(core.players()).toEqual(restoreCore.players());
  expect(core.stateHistory()).toEqual(restoreCore.stateHistory());
});
