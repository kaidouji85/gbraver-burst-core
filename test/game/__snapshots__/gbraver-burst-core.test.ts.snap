// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`初期状態を正しく作ることができる: initial-state 1`] = `
[
  {
    "activePlayerId": "player2",
    "effect": {
      "name": "StartGame",
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 1600,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
  {
    "activePlayerId": "player2",
    "effect": {
      "name": "InputCommand",
      "players": [
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 5,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "player1",
          "selectable": true,
        },
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 5,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "player2",
          "selectable": true,
        },
      ],
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 1600,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
]
`;

exports[`正しくゲームを進めることができる: progress 1`] = `
[
  {
    "activePlayerId": "player2",
    "effect": {
      "attacker": "player2",
      "attackerBattery": 2,
      "defenderBattery": 3,
      "name": "BatteryDeclaration",
      "originalBatteryOfAttacker": 2,
      "originalBatteryOfDefender": 3,
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 1600,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 3,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
  {
    "activePlayerId": "player2",
    "effect": {
      "attacker": "player2",
      "isDeath": false,
      "name": "Battle",
      "result": {
        "name": "Miss",
      },
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 1600,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 3,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
  {
    "activePlayerId": "player2",
    "effect": {
      "battleResult": {
        "name": "Miss",
      },
      "defender": "player1",
      "name": "RightItself",
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 1600,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 3,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
  {
    "activePlayerId": "player2",
    "effect": {
      "endArmdozerEffects": [],
      "name": "UpdateRemainingTurn",
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 1600,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 3,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
  {
    "activePlayerId": "player1",
    "effect": {
      "name": "TurnChange",
      "reason": "Normal",
      "recoverBattery": 3,
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 1600,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 3,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
  {
    "activePlayerId": "player1",
    "effect": {
      "name": "InputCommand",
      "players": [
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 5,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "player1",
          "selectable": true,
        },
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "player2",
          "selectable": true,
        },
      ],
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 1600,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 3,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
]
`;
